### ISSUE TITLE
feature request: entity match filter

### ISSUE THREAD

**Issue by mrbig** [oroinc/crm#328]
Dear OroCRM developers,

We've realized that there are use cases where the curren't entity filters aren't capable to create the exact list we want.

As an example let's take a scheme where you have students and their exams:

```
student
   id
   name

exam
  id
  student_id
  date
  grade
```

(Note: students could be contacts, but for the example I will use this term because I find it more descriptive)

Currently the relational filters are implemented as an EXIST query, which works nice when there are missing items on the right side of the relation.

So a filter for students who had an exam with grade A looks something like this:

```
SELECT ... FROM student WHERE
   EXISTS (SELECT ... FROM exam WHERE grade = 'A')
```
A filter for students who had an exam in 2018:
```
SELECT ... FROM student WHERE
   EXISTS (SELECT ... FROM exam WHERE date BETWEEN '2018-01-01' AND '2018-12-31')
```

Now I someone wants to make a list of the students, who got grade A in 2018 then the filter will create something like this:

```
SELECT ... FROM student WHERE
   EXISTS (SELECT ... FROM exam WHERE grade = 'A')
   AND EXISTS (SELECT ... FROM exam WHERE date BETWEEN '2018-01-01' AND '2018-12-31')
```

But this will also match students who had BOTH an exam in 2017 with grade A, and an exam in 2018 with any kind of grade.

To solve this issue instead of modifying the current filter behavior we would suggest a new operator for this kind of filtering. The very same problem was solved by MongoDB with the $elemMatch operator https://docs.mongodb.com/manual/reference/operator/query/elemMatch/

OroCRM could have something like that also: creating a new filter type, something like "related entity filter". The UI could be alike to the Conditions Group filter, with a subtle exception: the user should select a related entity first, then the filters couldbe added by dragging them to the box.

In our example by adding such a filter and selecting student -> exam referenced entity would create an query like this:

```
SELECT ... FROM student WHERE
   EXISTS (
      SELECT ... FROM exam WHERE
         grade = 'A'
         AND date BETWEEN '2018-01-01' AND '2018-12-31'
   )
```

We believe this would solve several use cases with filters, and would make OroCRM a more great product.

**Comment by mbessolov** [comment]
Hello @mrbig 

Thank you for the detailed report - it's actually better than our own description in the internal ticket.
Yes, we do have this feature as one of the possible improvements to the reporting engine in the product roadmap. No ETA.

**Comment by anyt** [comment]
Implemented in 3.1.7 version, see https://github.com/oroinc/platform/commit/70071776ed70ee4eebe92ca894eddbde35173a99.

**Comment by mrbig** [comment]
For others struggling to use this feature here is some hint:
 - it is not enabled by default, you have to tick the "Group same-entity conditions within condition groups" option under the Display settings
 - after that if a condition group's filters refer to the same entity, then the an referenced entity must satisfy all the filters in the condition group for the group to succeed.
